src/challenge/hooks/useChallengeLoader.tsx

```diff
@@ -26,14 +26,29 @@ const useChallengeLoader = (props: HookProps) => {
   const [guideMd, setGuideMd] = useState<string>(
     "*Loading the guide... Please wait*"
   );
+  const [isLoadingGuide, setIsLoadingGuide] = useState<boolean>(true);
   const [starterCode, setStarterCode] = useState<string | undefined>(undefined);
   const [savedCode, setSavedCode] = useState<string | undefined>(undefined);
+  const [isLoadingCode, setIsLoadingCode] = useState<boolean>(true);
   const [additionalFilesLoaded, setAdditionalFilesLoaded] =
     useState<AdditionalFilesContents>({});
 
   const authContext = useContext(SessionContext);
+  const authContextRef = useRef<SessionContextType>(authContext);
   const additionalFilesLoadedRef = useRef<AdditionalFilesContents>({});
 
+  const codePathRef = useRef<string>(props.codePath);
+  const guidePathRef = useRef<string>(props.guidePath);
+
+  useEffect(() => {
+    authContextRef.current = authContext;
+  }, [authContext]);
+
+  useEffect(() => {
+    codePathRef.current = props.codePath;
+    guidePathRef.current = props.guidePath;
+  }, [props.codePath, props.guidePath]);
```

Comment:
some juggling here to reduce race conditions. By stashing stuff from the props in refs it means that anyone who wants to read these values can do so without an explicit dependency, and they are guaranteed to see the latest value at all times.